feat: keyless wallet & apple google oauth login

sidmorizon opened the pull request.

franco-chan committed:
onekeyid login page

franco-chan committed:
social & otp

sidmorizon committed:
refactor: rename setMainWindow function for clarity and update related references

- Renamed `setMainWindow` to `setMainWindowForOAuthServer` in `oauthLocalServer` for better clarity.
- Updated all references in `app.ts` to reflect the new function name.
- Enhanced the `BaseSkeleton` component to accept a forwarded ref, improving its integration with parent components.
- Updated `HyperlinkText` to use a lazy-loaded default internationalization instance.
- Improved OAuth state handling in `openOAuthPopupWeb` and `useSupabaseAuth` for enhanced security and reliability.

sidmorizon committed:
feat: enhance OAuth flow with oneKeyState handling and refactor related components

- Added support for `oneKeyState` in the OAuth callback handling to improve security and state validation.
- Updated `openOAuthPopupDesktopLocalhost` to manage OAuth sessions effectively, including session persistence options.
- Refactored `useSupabaseAuth` to streamline OAuth processes and ensure compatibility with the new state handling.
- Removed the deprecated `openOAuthPopupDesktopLocalhost` function to clean up the codebase.
- Enhanced error handling across various OAuth methods to provide clearer feedback on authentication issues.

sidmorizon committed:
refactor: update OAuth callback paths and enhance state handling

- Changed OAuth callback paths for desktop and web to `/oauth_callback_desktop` and `/oauth_callback_web`, respectively, improving clarity and consistency.
- Introduced `ensureOneKeyOAuthState` utility to guarantee the presence of the `ONEKEY_OAUTH_STATE_KEY` parameter in redirect URLs, enhancing security.
- Updated various components to utilize the new callback paths and state handling, ensuring robust OAuth flow and validation.
- Improved error handling and state validation in the OAuth process to mitigate potential security risks.

sidmorizon committed:
refactor: update OAuth server to use system-assigned ports and improve callback handling

- Removed fixed port range for OAuth callbacks, allowing the server to listen on a dynamically assigned port.
- Enhanced the OAuth callback handling logic to improve error responses and streamline the process of receiving authorization codes.
- Updated documentation to reflect changes in the OAuth flow and the new method of handling redirect URLs.
- Improved error messages for better user feedback during OAuth server startup.

sidmorizon committed:
refactor: streamline OAuth flow in Chrome extension

- Introduced a new `getRedirectUrl` function to standardize the retrieval of the OAuth redirect URL, ensuring it matches Google Cloud Console configuration.
- Updated `openOAuthPopupExtIdentity` to utilize the new redirect URL function and accept an optional `authUrl` parameter for improved flexibility.
- Refactored the OAuth flow processing to simplify the handling of authentication URLs and session management.
- Enhanced `useSupabaseAuth` to support the new `authUrl` parameter, aligning the extension's OAuth handling with web standards.

franco-chan committed:
ui

sidmorizon committed:
refactor: consolidate OAuth handling across platforms

- Removed OAuth2 configuration from Chrome manifest files, streamlining the setup process.
- Simplified the `IHandleOAuthSessionPersistenceParams` type by removing unnecessary properties.
- Introduced a unified `OAuthPopup` class structure for handling OAuth across web, desktop, extension, and native platforms.
- Enhanced `useSupabaseAuth` to leverage the new `OAuthPopup` implementation, improving session management and code clarity.
- Updated various components to ensure compatibility with the new OAuth structure and improve overall maintainability.

sidmorizon committed:
refactor: remove obsolete OAuth components and types

- Deleted unused OAuth handling files for desktop, extension, native, and web platforms to streamline the codebase.
- Removed related types and utility functions that are no longer necessary, enhancing maintainability and clarity.
- This cleanup aligns with recent refactoring efforts to consolidate OAuth handling across platforms.

revan-zhang commented:
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-critical.svg) Critical | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-high.svg) High | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-medium.svg) Medium | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-low.svg) Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/rayston92/pr-checks/f729e3c8-6bcf-4c70-90b3-a6d6ee93d052?refs=39ec34e7d227c12ad45ba1c64a9fcb3b61815a00&source=prChecksComment) |
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/rayston92/pr-checks/f729e3c8-6bcf-4c70-90b3-a6d6ee93d052/license?refs=39ec34e7d227c12ad45ba1c64a9fcb3b61815a00&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

sidmorizon committed:
fix: google sign in for iOS

sidmorizon committed:
fix: ios google login

sidmorizon committed:
fix: upgrade google sign in sdk

sidmorizon committed:
fix: lint

sidmorizon committed:
feat: add Apple Sign-In setup guide

- Introduced a comprehensive guide for configuring Apple Sign-In with Supabase for the OneKey app, focusing on web platform setup.
- Documented prerequisites, local development testing, and detailed steps for Apple Developer Console and Supabase configuration.
- Included troubleshooting tips and security considerations to enhance user understanding and implementation of Apple Sign-In.
- Updated OAuthPopup components to improve callback URL validation and nonce generation for enhanced security during authentication.

franco-chan committed:
fix conflix

sidmorizon committed:
apple login

sidmorizon committed:
fix(crypto): update crypto.subtle.digest test result

sidmorizon commented on packages/kit/src/components/OneKeyAuth/OAuthPopup/APPLE_SIGNIN_SETUP.md:
删除

sidmorizon committed:
fix: lint

sidmorizon committed:
fix: lint

sidmorizon committed:
apple login

sidmorizon committed:
fix: add ios apple sign in entitlements

franco-chan committed:
ui

sidmorizon committed:
apple native login

franco-chan committed:
ui

franco-chan committed:
Merge branch 'keyless-wallet-2' into fix/android-google-login

sidmorizon committed:
Merge branch 'fix/android-google-login' of github.com:OneKeyHQ/app-monorepo into fix/android-google-login

sidmorizon committed:
feat: add Keyless Wallet V2 functionality and update onboarding flow

franco-chan committed:
i18n

sidmorizon committed:
feat: enhance Keyless Wallet V2 with social login support and onboarding improvements

sidmorizon committed:
fix: merge

sidmorizon committed:
fix: lint

franco-chan committed:
responsive design

franco-chan committed:
Merge branch 'fix/android-google-login' of github.com:OneKeyHQ/app-monorepo into fix/android-google-login

franco-chan committed:
i18n & responsive design

sidmorizon committed:
feat: enhance Keyless Wallet V2 with new onboarding features and debug panel for custom mnemonic management

sidmorizon committed:
feat: refactor PinInputLayout to support forward ref and enhance focus handling; update KeylessOnboardingDebugPanel button text; clean up unused code in OneKeyIDLoginPage and VerifyPinPage

sidmorizon committed:
Merge branch 'fix/android-google-login' of github.com:OneKeyHQ/app-monorepo into fix/android-google-login

# Conflicts:
#	packages/kit/src/views/Onboardingv2/components/PinInputLayout.tsx

sidmorizon committed:
fix: update onVerifyRes type to support Promise and ensure consistent async handling in PasswordVerifyContainer

sidmorizon committed:
refactor: streamline keyless wallet handling by replacing utility functions with direct property access; update onboarding timeout messages for clarity and security; remove unused PinInputLayout2 component

sidmorizon committed:
feat: add useKeylessWalletExistsLocal hook to check local keyless wallet existence; update related components to utilize new functionality

sidmorizon committed:
feat: implement Reset PIN flow for Keyless Wallet, including navigation updates and secure storage for mnemonic password; enhance onboarding with new actions and error handling

sidmorizon committed:
feat: enhance ServiceKeylessWallet to include juiceboxShareX and backendShareX for improved share recovery; update related types and logic for handling shares

sidmorizon committed:
feat: update Keyless Wallet onboarding flow with new modes for Reset PIN and Verify PIN; refactor related components for improved navigation and state handling

sidmorizon committed:
feat: improve Keyless Wallet onboarding by enhancing PIN verification dialog and refactoring login state management; update error handling in ServiceKeylessWallet for better robustness

sidmorizon committed:
refactor: rename functions and update logic in Keyless Wallet service for clarity; enhance onboarding flow by checking wallet creation status on the server

franco-chan committed:
animations & account selector

franco-chan committed:
fix badge font size

sidmorizon committed:
feat: enhance Keyless Wallet onboarding flow by adding refresh token handling and improving navigation logic; refactor related components for better state management and user experience

sidmorizon committed:
feat: add error handling for keyless wallet retrieval and refresh token processes; improve navigation logic in onboarding flow

sidmorizon committed:
Merge branch 'fix/android-google-login' of github.com:OneKeyHQ/app-monorepo into fix/android-google-login

sidmorizon committed:
refactor: remove socialLoginProvider prop from WalletAvatarProps and derive it from wallet keyless details for improved clarity and maintainability

sidmorizon committed:
refactor: adjust wallet creation flow in ServiceKeylessWallet to ensure juiceboxShare is uploaded before backend share, preventing concurrent wallet creation

sidmorizon committed:
fix: merge conlict

sidmorizon committed:
fix: update Google OAuth client IDs in Info.plist and authConsts.ts for iOS compatibility

sidmorizon committed:
feat: add display of Google OAuth client IDs in OneKeyIDApiTests for enhanced debugging

sidmorizon committed:
feat: integrate Juicebox for Keyless Wallet discovery and recovery

sidmorizon committed:
feat: implement juicebox api calls and optimize pin verification error handling

sidmorizon committed:
fix: lint

sidmorizon committed:
fix: improve error handling in ServiceKeylessWallet for backend share parsing

sidmorizon committed:
chore: add TODO comments for future enhancements in JuiceboxClient and keylessWalletTypes

franco-chan committed:
authentication page improvement: light & dark mode styles and layout

sidmorizon committed:
feat: add juicebox-sdk integration and keyless wallet reset debug feature

sidmorizon committed:
fix(kit): support google login in onboarding get started page

sidmorizon committed:
fix: owner id

(cherry picked from commit 06ee0c4386f1d85ac045174da034380f44dacc44)

originalix committed:
fix: add pre-install hook for GitHub URL configuration

originalix committed:
fix: move pre-install hook to package.json for EAS build configuration

originalix committed:
Update provisioning profiles and add JuiceboxSdk dependency

originalix committed:
chore: xcode profile

originalix committed:
fix: update code signing settings and resolve JuiceboxSdkFfi linking issues

sidmorizon committed:
fix(kit): add error toast for OAuth sign-in failures

(cherry picked from commit 425b31cad2bfc3bfd9658e71526c44b55e5afec4)

sidmorizon committed:
fix: merge

sidmorizon committed:
fix(auth): fix iOS Google OAuth nonce handling and reorganize auth constants

- Return undefined nonce for iOS Google Sign-In to fix authentication issue
- Add improved comments explaining nonce behavior in OAuth flow
- Reorganize Supabase and Keyless config sections with clear separators

sidmorizon committed:
Merge branch 'x' into fix/android-google-login

# Conflicts:
#	apps/mobile/ios/Podfile.lock
#	apps/mobile/package.json

sidmorizon committed:
fix: build

sidmorizon committed:
Merge branch 'x' into fix/android-google-login

limichange commented on packages/kit-bg/src/services/ServiceKeylessWallet/ServiceKeylessWallet.ts:
```diff
@@ -1072,6 +1108,610 @@ class ServiceKeylessWallet extends ServiceBase {
+
+    if (isSuccess && backendShareStr) {
+      try {
+        const result = JSON.parse(backendShareStr) as
+          | IKeylessBackendShare
+          | undefined;
+        if (result) {
+          return result;
+        }
+      } catch (e) {
+        throw new OneKeyLocalError('Failed to parse keyless backend share');
```
**[P1] JWT Token Not Validated Before Trust**

The `buildKeylessOwnerIdFromSocialToken` function decodes and trusts the JWT token content without verifying its signature. While this token comes from Supabase OAuth, parsing it client-side without validation creates a risk if the token is tampered with in transit or stored insecurely.

**Improvement suggestion**: Consider validating the JWT signature before extracting claims, or ensure the token is always validated server-side before any sensitive operations.

limichange commented on packages/kit-bg/src/services/ServiceKeylessWallet/ServiceKeylessWallet.ts:
```diff
@@ -1072,6 +1108,610 @@ class ServiceKeylessWallet extends ServiceBase {
+    token: string;
+    pin: string;
+  }): Promise<IKeylessJuiceboxShare | null> {
+    const { token, pin } = params;
+
+    if (!token) {
+      throw new OneKeyLocalError('Missing token');
+    }
+
+    if (!pin) {
+      throw new OneKeyLocalError('Missing pin');
+    }
+
+    const ownerId = this.buildKeylessOwnerIdFromSocialToken({ token });
+
+    const { JuiceboxClient } = await import('./utils/JuiceboxClient');
+    const juiceboxClient = new JuiceboxClient();
+    await juiceboxClient.exchangeToken(token);
+    try {
+      const secret = await juiceboxClient.recover({
+        pin,
+        userInfo: ownerId,
+      });
+
+      const parts = secret.split('--');
+      const backendShareXStr = parts.pop();
+      if (!backendShareXStr) {
+        throw new OneKeyLocalError(
+          'Failed to get keyless juicebox share: backendShareXStr is empty',
+        );
+      }
+      const backendShareX = parseInt(backendShareXStr || '0', 10);
+      const juiceboxShare = parts.join('');
+      if (!juiceboxShare) {
```
**[P1] Sensitive Data in Memory Without Secure Cleanup**

The `createKeylessWalletToServer` function generates and handles sensitive cryptographic material (mnemonic, mnemonicPasswordBytes, shares) in memory. These values are not explicitly zeroed after use, which could leave sensitive data accessible in memory for an extended period.

**Improvement suggestion**: Consider implementing secure memory cleanup by zeroing out sensitive byte arrays (mnemonicPasswordBytes, mnemonicPasswordShares) after use using a secure wipe function.

limichange commented on packages/kit-bg/src/services/ServiceKeylessWallet/utils/JuiceboxClient.ts:
```diff
@@ -0,0 +1,231 @@
+    }
+  }
+
+  /**
```
**[P1] Global Callback Function Assignment**

Setting `globalThis.JuiceboxGetAuthToken` as a global callback function creates potential security risks:
1. Other code could potentially override this callback
2. The callback persists beyond its intended scope
3. No validation that the callback is being called from expected context

**Improvement suggestion**: Consider using a more secure pattern like passing the auth callback directly to the SDK if supported, or implementing additional validation in the callback to verify the caller.

limichange commented on packages/kit-bg/src/services/ServiceKeylessWallet/ServiceKeylessWallet.ts:
```diff
@@ -1072,6 +1108,610 @@ class ServiceKeylessWallet extends ServiceBase {
+      const backendShareX = parseInt(backendShareXStr || '0', 10);
+      const juiceboxShare = parts.join('');
+      if (!juiceboxShare) {
+        throw new OneKeyLocalError(
```
**[P2] Error Handling TODO for Critical Flow**

There's a TODO comment `// TODO cleanup if error occurs` at the end of `createKeylessWalletToServer`. If uploading to juicebox or backend fails partway, the wallet state could be left inconsistent - shares may be partially uploaded without rollback.

**Improvement suggestion**: Implement proper error handling with rollback mechanism. If backend share upload fails after juicebox share is uploaded, consider deleting the juicebox share to maintain consistency.

limichange commented on packages/kit-bg/src/services/ServiceKeylessWallet/utils/keylessMnemonicPasswordStorage.ts:
```diff
@@ -0,0 +1,133 @@
+  );
+
+  // Convert hex to base64 for storage
+  const encryptedPayloadBase64 = bufferUtils.bytesToBase64(
+    bufferUtils.hexToBytes(encryptedPayloadHex),
+  );
+
+  // 3. Store encrypted data, prefer secureStorage if available
+  await storageSetItem(key, encryptedPayloadBase64);
+}
+
+/**
+ * Get mnemonicPassword from local storage and decrypt it.
+ */
+async function getMnemonicPasswordFromStorage(params: {
+  ownerId: string;
+  backgroundApi: IBackgroundApi;
+}): Promise<string | null> {
+  const { ownerId, backgroundApi } = params;
+
+  // 1. Build unique key for this ownerId
```
**[P2] Double Encryption with Same Key Source**

The mnemonicPassword is encrypted using `buildKeylessLocalEncryptionKey` and stored. The encryption key is derived from the passcode, but if an attacker gains access to the encrypted storage and the passcode (e.g., through a device compromise), they can decrypt the mnemonicPassword and potentially recover the wallet.

**Improvement suggestion**: Consider using additional entropy sources (device-specific keys, biometrics) to strengthen the encryption key derivation for this sensitive data.

limichange reviewed (commented):
## Code Review Report

| File | Line | Issue |
|------|------|------|
| packages/kit-bg/src/services/ServiceKeylessWallet/ServiceKeylessWallet.ts | 1176-1183 | P1 JWT Token Not Validated Before Trust |
| packages/kit-bg/src/services/ServiceKeylessWallet/ServiceKeylessWallet.ts | 1254-1284 | P1 Sensitive Data in Memory Without Secure Cleanup |
| packages/kit-bg/src/services/ServiceKeylessWallet/utils/JuiceboxClient.ts | 101 | P1 Global Callback Function Assignment |
| packages/kit-bg/src/services/ServiceKeylessWallet/ServiceKeylessWallet.ts | 1285 | P2 Error Handling TODO for Critical Flow |
| packages/kit-bg/src/services/ServiceKeylessWallet/utils/keylessMnemonicPasswordStorage.ts | 66-83 | P2 Double Encryption with Same Key Source |

Found 5 issue(s).

sidmorizon committed:
Merge branch 'x' into fix/android-google-login

sidmorizon committed:
Merge branch 'x' into fix/android-google-login
